### PR TITLE
Latency tests analyze integration

### PR DESCRIPTION
**Pull request by Paullgk**
This PR adds a first integration of latency tests analysis: publisher/subscriber timestamp data analysis, statistics and graph generation, integration into final CI report.
To avoid any regression, on CI runtime, the latency test analyze is done in a separate script `compile_latency.sh`, different from main `compile.sh` script. The script can be executed by calling cqfd with `build_latency` flavor.

Following files are added:

- `compile_latency.sh`: main core script. Call python script for data analysis, statistics and graph generation. Finally, do the integration into final CI report.
- `report-generator/timestamp_analysis/post_processing_utils.py` and r`eport-generator/timestamp_analysis/timestamp_analysis.py`: analysis of timestamp data files, and generation of stats/graph belonging to them.

### CODE REVIEW

File: report-generator/compile_latency.sh [hunk 1]
```diff
@@ -0,0 +1,53 @@
+#!/bin/bash
+
+
```
Comment (deribaucourt) [inline report-generator/compile_latency.sh:4]: set -e

File: report-generator/compile_latency.sh [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+{
+    # Generate graph and stats associated to a publisher and subscriber test
+    # Parameters: $1: publisher result file
+    # $2: subscriber result file
+
+    local PUB_DATA_FILE=$1
+    local SUB_DATA_FILE=$2
+    python3 timestamp_analysis/timestamp_analysis.py \
+    $PUB_DATA_FILE \
+    $SUB_DATA_FILE >> $STATS_FILE
```
Comment (deribaucourt) [inline report-generator/compile_latency.sh:36]: rm $STATS_FILE at the beginning of the script to avoid importing previous aborted tests

File: report-generator/compile_latency.sh [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+}
+
+
+add_latency_results()
+{
+    # Generate and add latency tests results to final report
+    # Parameter: $1: path to tests results directory
+    echo "==  Latency tests results ==" >> "$LATENCY_ADOC_FILE"
```
Comment (deribaucourt) [inline report-generator/compile_latency.sh:44]: rm LATENCY_ADOC_FILE to avoid importing previous aborted tests

File: report-generator/compile_latency.sh [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+        sed -i 's/$/ +/' $STATS_FILE # Add '+' character at the end of each file sub stat line for making carriage return
+        cat $STATS_FILE >> "$LATENCY_ADOC_FILE" # Add sub stats data in report file
+        rm "$STATS_FILE"
+        echo " " >> "$LATENCY_ADOC_FILE"
+    done
+
+
+}
+
+add_latency_results $1
```
Comment (deribaucourt) [inline report-generator/compile_latency.sh:70]: Check if $1 was provided, exit with an help message if not.

File: report-generator/compile_latency.sh [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+
+add_latency_results()
+{
+    # Generate and add latency tests results to final report
+    # Parameter: $1: path to tests results directory
+    echo "==  Latency tests results ==" >> "$LATENCY_ADOC_FILE"
+    local TESTS_RESULTS_DIR=$1
+    local SUB_RESULTS_FILES_HYPERVISOR="$(ls "$TESTS_RESULTS_DIR"sub*hypervisor*)" # Getting all sub files results
+
+    local PUB_RESULT_FILE_HYPERVISOR=$TESTS_RESULTS_DIR"pub_results_hypervisor_standalone" # Publisher file result
```
Comment (deribaucourt) [inline report-generator/compile_latency.sh]: `"${TESTS_RESULTS_DIR}/pub_results_hypervisor_standalone"`

File: report-generator/compile_latency.sh [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+}
+
+
+add_latency_results()
+{
+    # Generate and add latency tests results to final report
+    # Parameter: $1: path to tests results directory
+    echo "==  Latency tests results ==" >> "$LATENCY_ADOC_FILE"
+    local TESTS_RESULTS_DIR=$1
+    local SUB_RESULTS_FILES_HYPERVISOR="$(ls "$TESTS_RESULTS_DIR"sub*hypervisor*)" # Getting all sub files results
```
Comment (deribaucourt) [inline report-generator/compile_latency.sh]: Syntax looks weird. Try something simpler:

`SUB_RESULTS_FILES_HYPERVISOR="${TESTS_RESULTS_DIR}/sub*hypervisor*"`

File: report-generator/compile_latency.sh [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+
+    local PUB_RESULT_FILE_HYPERVISOR=$TESTS_RESULTS_DIR"pub_results_hypervisor_standalone" # Publisher file result
+
+    # Compute hypervisor subscriber tests results
+    for SUB_RESULT_FILE in $SUB_RESULTS_FILES_HYPERVISOR # For each sub result file
```
Comment (deribaucourt) [inline report-generator/compile_latency.sh:50]: You don't need to comment each line. It's fine but not necessary. You're code is clear and self explanatory :slightly_smiling_face:

File: report-generator/compile_latency.sh [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+    do
+        generate_latency_results $PUB_RESULT_FILE_HYPERVISOR $SUB_RESULT_FILE # Generate graph and stats associate to it
+        local SUB_MACHINE_NAME="$(echo $SUB_RESULT_FILE|cut -d'/' -f4|cut -d'_' -f4)" # Getting sub hypervisor machine name
```
Comment (deribaucourt) [inline report-generator/compile_latency.sh:53]: Add a comment showing the before/after transformation of complex cuts/grep/awk/sed expressions for clarity. Ex:
`# a/b/c d/f/e --> f`

File: report-generator/compile_latency.sh [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+        echo "=== $SUB_MACHINE_NAME" >> "$LATENCY_ADOC_FILE" # Subtitle for each subscriber
+        echo "image::./include/sub_"$SUB_MACHINE_NAME"_delay.png[]" >> "$LATENCY_ADOC_FILE" # Include delay graph
+        echo "image::./include/pub_standalone_interval_between.png[]" >> "$LATENCY_ADOC_FILE" # Include time between packets graph
```
Comment (deribaucourt) [inline report-generator/compile_latency.sh:58]: So this image will be duplicated in the doc for each SUB?

File: report-generator/compile_latency.sh [hunk 10]
```diff
@@ -0,0 +1,53 @@
...
+        echo " " >> "$LATENCY_ADOC_FILE"
+
+        sed -i 's/$/ +/' $STATS_FILE # Add '+' character at the end of each file sub stat line for making carriage return
+        cat $STATS_FILE >> "$LATENCY_ADOC_FILE" # Add sub stats data in report file
```
Comment (deribaucourt) [inline report-generator/compile_latency.sh:62]: isn't there instead an adoc function to include a raw text file?

File: report-generator/timestamp_analysis/post_processing_utils.py [hunk 11]
```diff
@@ -0,0 +1,48 @@
+# Copyright (C) 2023, RTE (http://www.rte-france.com)
```
Comment (deribaucourt) [inline report-generator/timestamp_analysis/post_processing_utils.py:1]: This is Thibault's script right?
If that's the case:
1. set him as author with `git commit --author "Thibault X <thibault.x@mail.com>"`
2. clearly state in the commit message that this commit was imported and if any modifications have been done, move them in a separate commit.

File: report-generator/.cqfd/docker/Dockerfile [hunk 12]
```diff
@@ -10,10 +10,13 @@ RUN apt-get update && apt-get install -y --no-install-recommends \
 	asciidoc \
 	bash \
 	locales \
+	python3-pip \
```
Comment (deribaucourt) [inline report-generator/.cqfd/docker/Dockerfile]: Don't you prefer installing python3-matplotlib directly?
We don't use pip afterwards.

File: report-generator/.cqfd/docker/Dockerfile [hunk 13]
```diff
@@ -10,10 +10,13 @@ RUN apt-get update && apt-get install -y --no-install-recommends \
...
 	ruby \
 	ruby-asciidoctor-pdf \
 	xmlstarlet \
 	&& rm -rf /var/lib/apt/lists/*
 
 RUN sed -i "s/# en_US\.UTF-8 UTF-8/en_US\.UTF-8 UTF-8/" /etc/locale.gen
 RUN locale-gen
+
+RUN pip3 install matplotlib
```
Comment (deribaucourt) [inline report-generator/.cqfd/docker/Dockerfile]: Should be moved in the apt RUN statement if conserved to prevent creating temporary docker images (docker creates snapshots for each statement).

**Review by deribaucourt — CHANGES_REQUESTED** [review]

File: report-generator/compile_latency.sh (hunk 3 above)
Comment (Paullgk) [reply report-generator/compile_latency.sh:25]: This is already done, see the last lines of the script

File: report-generator/compile_latency.sh (hunk 6 above)
Comment (Paullgk) [reply report-generator/compile_latency.sh]: Yes, I need to perform a `ls` to list all files with pattern `sub` and `hypervisor`. Without it, it will not work.

File: report-generator/compile_latency.sh (hunk 10 above)
Comment (Paullgk) [reply report-generator/compile_latency.sh:62]: Yes, we could use asciidoc `include` function, but as ascii doc pdf conversion is done after deleting stat file, it is much simple to use cat before removing it.

File: report-generator/compile_latency.sh (hunk 9 above)
Comment (Paullgk) [reply report-generator/compile_latency.sh:58]: Yes, as it is a graph of delay between each packets sent by publisher. It will be improve in future to be display only one time.

File: report-generator/.cqfd/docker/Dockerfile [hunk 14]
```diff
@@ -10,10 +10,13 @@ RUN apt-get update && apt-get install -y --no-install-recommends \
...
 	locales \
+	python3-matplotlib \
 	ruby \
 	ruby-asciidoctor-pdf \
 	xmlstarlet \
 	&& rm -rf /var/lib/apt/lists/*
 
 RUN sed -i "s/# en_US\.UTF-8 UTF-8/en_US\.UTF-8 UTF-8/" /etc/locale.gen
 RUN locale-gen
+
```
Comment (deribaucourt) [inline report-generator/.cqfd/docker/Dockerfile]: configure your editor to put exactly one empty newline at the end of files.

**Review by deribaucourt — DISMISSED** [review]

**Review by eroussy — APPROVED** [review]
